### PR TITLE
Align rope of chatglm with llama

### PR DESCRIPTION
**Pull request by hkvision**
This PR tries to use the same logic of llama rope for chatglm, which is the first step for use fuse rope xpu.

Since rotary_pos_emb is already indexed by position_id (i.e. cos/sin) before passing to self-attention forward, more work needs to be done followed by this PR to align with llama (where cos/sin are calculated in the kernel and then indexed by position_id instead).

Any comments on this? Not sure after such modification there would still be performance benefits. @jason-dai @yangw1234

### CODE REVIEW

**Comment by hkvision** [comment]
Updated, the current implementation uses the existing xpu fuse rope kernel to the largest extent.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 1]
```diff
@@ -119,19 +119,56 @@ def chatglm2_attention_forward_8eb45c(
...
+            key_layer = apply_rotary_pos_emb_chatglm(key_layer, rotary_pos_emb)
 
     if self.multi_query_attention:
         key_length = key_layer.size(0)
         query_group_size = self.num_attention_heads_per_partition // \
             self.num_multi_query_groups_per_partition
+        # if rotary_pos_emb is not None and use_fuse_rope:
+        #     key_layer = key_layer.unsqueeze(-3)
+        # else:
+        #     key_layer = key_layer.permute(1, 2, 0, 3).unsqueeze(-3)  # [bs, nh/k, sl, hn]
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:168]: In theory this permute can be merged with xpu kernel's output shape, but don't know why it gets the error:
```
  File "/home/arda/anaconda3/envs/kai-llm-pip/lib/python3.9/site-packages/bigdl/llm/transformers/models/chatglm2.py", line 193, in chatglm2_attention_forward_8eb45c
    key_layer = key_layer.contiguous().view((batch_size,
RuntimeError: shape '[1, 32, 1, 128]' is invalid for input of size 53248
```

**Comment by hkvision** [comment]
Result update and analysis is here: https://github.com/analytics-zoo/nano/issues/608#issuecomment-1755356079

**Comment by hkvision** [comment]
TODO: may need to further compare the result after rotary embedding.
Seems the generated texts are different before/after the change, but both are meaningful and in good quality.

File: python/llm/src/bigdl/llm/transformers/models/chatglm2.py [hunk 2]
```diff
@@ -187,7 +275,8 @@ def chatglm2_attention_forward_8eb45c(
     # ==================================
     # core attention computation
     # ==================================
-
+    # print("----query", query_layer.shape)
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/chatglm2.py:278]: clear these

**Comment by yangw1234** [comment]
rms_norm seems to change the result a bit for llama. See https://github.com/intel-analytics/BigDL/pull/9081

**Comment by hkvision** [comment]
> rms_norm seems to change the result a bit for llama. See #9081
> 
> I guess it should be an alarm if the first a few output tokens are different.

Sure, but seems chatglm2 changes output after modifying rope now, will further check this.

**Comment by hkvision** [comment]
Pending for fuse, @sgwhat will break fuse rms into a separate PR.

**Comment by hkvision** [comment]
https://github.com/intel-analytics/BigDL/pull/9350 follow-up PR, closing this one.

**Pull request closed without merging**